modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/QuantumToolkit.java

```diff
@@ -1497,44 +1553,95 @@ public void run() {
+                        // +-----------+-----------+  .  +-------+
+
+                        // Walk through all same-size "M" tiles
+                        int xOffset = 0;
+                        int yOffset = 0;
+                        var mTileWidth = tileWidth;
+                        var mTileHeight = tileHeight;
+                        while (mTileWidth == tileWidth && xOffset < w) {
+                            yOffset = 0;
+                            mTileHeight = tileHeight;
+                            while (mTileHeight == tileHeight && yOffset < h) {
+                                renderTile(x, xOffset, y, yOffset, mTileWidth, mTileHeight,
+                                        buffer, rf, tileRttCache, pImage);
+                                yOffset += tileHeight;
+                                mTileHeight = Math.min(tileHeight, h - yOffset);
+                            }
+                            xOffset += tileWidth;
+                            mTileWidth = Math.min(tileWidth, w - xOffset);
+                        }
+                        // Walk through remaining same-height "R" tiles, if any
+                        int rOffset = 0;
+                        int rTileHeight = tileHeight;
+                        if (mTileWidth > 0) {
+                            while (rTileHeight == tileHeight && rOffset < h) {
+                                renderTile(x, xOffset, y, rOffset, mTileWidth, rTileHeight, buffer, rf, tileRttCache, pImage);
+                                rOffset += tileHeight;
+                                rTileHeight = Math.min(tileHeight, h - rOffset);
+                            }
+                        }
+                        // Walk through remaining same-width "B" tiles, if any
+                        int bOffset = 0;
+                        int bTileWidth = tileWidth;
+                        if (mTileHeight > 0) {
+                            while (bTileWidth == tileWidth && bOffset < w) {
+                                renderTile(x, bOffset, y, yOffset, bTileWidth, mTileHeight, buffer, rf, tileRttCache, pImage);
+                                bOffset += tileWidth;
+                                bTileWidth = Math.min(tileWidth, w - bOffset);
+                            }
+                        }
+                        // Render corner "C" tile if needed
+                        if (mTileWidth > 0 && mTileHeight > 0 && rOffset > 0 && bOffset > 0) {
+                            renderTile(x, bOffset, y, rOffset, bTileWidth, rTileHeight, buffer, rf, tileRttCache, pImage);
                         }
                     }
-
-                    rt.unlock();
-
+                    else {
+                        // The requested size for the screenshot fits max texture size,
+                        // so we can directly render it in the target image.
+                        renderWholeImage(x, y, w, h, rf, pImage);
+                    }
```

Comment:
I tried to write this code using for loop to make it little easy for reader to understand. This is just a suggestion, I leave it to you to whether replace it or not,

```
	int mTileWidth = computeTileSize(w, maxTextureSize);
	int mTileHeight = computeTileSize(h, maxTextureSize);
	IntBuffer buffer = IntBuffer.allocate(mTileWidth * mTileHeight);

	int mTileXOffset = 0;
	int mTileYOffset = 0;
	for (mTileXOffset = 0; (mTileXOffset + mTileWidth) <= w; mTileXOffset += mTileWidth) {
		for (mTileYOffset = 0; (mTileYOffset + mTileHeight) <= h; mTileYOffset += mTileHeight) {
			renderTile(x, mTileXOffset, y, mTileYOffset, mTileWidth, mTileHeight,
					buffer, rf, tileRttCache, pImage);
		}
	}

	int rTileXOffset = mTileXOffset;
	int rTileWidth = w - rTileXOffset;
	if (rTileWidth > 0) {
		for (int rTileYOffset = 0; (rTileYOffset + mTileHeight) <= h; rTileYOffset += mTileHeight) {
			renderTile(x, rTileXOffset, y, rTileYOffset, rTileWidth, mTileHeight,
					buffer, rf, tileRttCache, pImage);
		}
	}

	int bTileYOffset = mTileYOffset;
	int bTileHeight = h - bTileYOffset;
	if (bTileHeight > 0) {
		for (int bTileXOffset = 0; (bTileXOffset + mTileWidth) <= w; bTileXOffset += mTileWidth) {
			renderTile(x, bTileXOffset, y, bTileYOffset, mTileWidth, bTileHeight,
					buffer, rf, tileRttCache, pImage);
		}
	}

	if (rTileWidth > 0 &&  bTileHeight > 0) {
		renderTile(x, rTileXOffset, y, bTileYOffset, rTileWidth, bTileHeight,
				buffer, rf, tileRttCache, pImage);
	}
```

Reply:
Thanks for you suggestion.

The original code is still too fresh in my mind for me to really be able to say which version is easier to understand, so I'll let others with a fresh pair of eyes be the judge of that.
However I think your code is arguably more concise - and overall more elegant - so I'd be happy to swap my version for yours, provided it is considered at least as readable as the original.
WDYT?

Reply:
We'll leave it up to you, but I note that it will require retesting if you adopt the suggested change.